### DISCUSSION TITLE
Thermal Radiation

### DISCUSSION THREAD

**Discussion by makeclean** [idaholab/moose#17818] [Q&A General] [answered]
Hi, feel like im bombarding you guys today :). Im trying the new ray tracing based thermal radiation stuff, so I setup my geometry in Cubit as usual, with what I think are appropriate boundaries and so on. I've been trying to use my Cubit generated mesh, but I first hit an error:
```
*** ERROR ***
This action adds MeshGenerator objects and therefore only works with a MeshGeneratorMesh.
```
So I changed my mesh section from 
```
[Mesh]
  file = '../mesh/hot-cup.e'
[]
```
to 
```
 type = MeshGeneratorMesh
 [bob]
   type = FileMesh
   file = '../mesh/hot-pipes.e'
 []
```
But I get the error message;
```
*** ERROR ***
Task add_mesh_generator is not registered to build MooseMesh derived objects
```
I presume from this, I can't use a file based mesh? Looking at the tests that exist for this functionality, it seems that they only use internally generated mesh, just checking, is that right? I can't use an exeternal mesh?

**Comment by makeclean** [comment, marked as answer]
Ah, got it, I need to use 
```
[Mesh]
  [bob]
    type = FileMeshGenerator
    file = '../mesh/hot-pipes-radiation.e'
  []
[]


```

**Reply by makeclean** [reply]
Ok, almost there, my geometry is the following;
![Screenshot from 2021-05-13 10-38-52](https://user-images.githubusercontent.com/2439754/118108088-7a87e900-b3d7-11eb-9e4c-75736d9337a7.png)
I have the following boundaries (side sets) defined vol-1 to vol-9 which are the pipes in the middle, I have vol-10 which are the inner walls of the block. I have a further side set called outer-walls which is the exterior boundary. I have a ConvectiveHeatFlux boundary condition on the outside wall. 

My problem, is it seems I need to define an additional RayBC either reflection or kill on the green (vol-10) side set, but when I do via a RayBCs section, I get an error 
```
*** ERROR ***
The following error occurred in the object "ray_study_uo_cavity", of type "ViewFactorRayStudy".

Requires one and only one ViewFactorRayBC.

Stack frames: 12
```
Without it, I get the following error
```
*** ERROR ***
Ray on processor 0 and thread 0 failed to trace at line 2137

Don't know what to do with a Ray after it hit an external
boundary at point (x,y,z)=( 0.37651,   0.4125, -0.0639156)!

When hitting an external RayBC, a Ray must either:
  Be killed by a RayBC
  Have its trajectory changed by the RayBC
by at least one of the executed RayBCs.

You need to either:
  Kill/change the Ray sooner with RayKernels, internal RayBCs, or a max distance
  Kill/change the Ray on the boundary with a RayBC

Boundaries that did not have any RayBCs:
  12 (outer-walls)
```
My diffuse radiaiton block looks like this;
```
[GrayDiffuseRadiation]
  [cavity]
    boundary = 'vol-1 vol-2 vol-3 vol-4 vol-5 vol-6 vol-7 vol-8 vol-9 vol-10'
    emissivity = '0.9 0.9 0.9 0.9 0.9 0.9 0.9 0.9 0.9 0.9'
    n_patches = '10 10 10 10 10 10 10 10 10 10'
    partitioners = 'centroid centroid centroid centroid centroid centroid centroid centroid centroid centroid'
    centroid_partitioner_directions = 'radial radial radial radial radial radial radial radial radial radial'
    temperature = temperature
    ray_tracing_face_type = 'gauss'
    view_factor_calculator = 'ray_tracing'
  []
[]
```
So I guess my question, has a few bits, is my setup sensible, and if so, what should I do about the outer-walls?

**Reply by loganharbour** [reply]
The orientation of the internal sidesets for this capability are _very_ important. Recall that for interior sidesets, the sideset will be associated with the side of an element, so there are two possible orientations. 

I suspect that what is happening here is that we're tracing rays from vol-10 outward to outer-walls because the orientation is off here, which is definitely not what you want to do. We trace rays based on the half space of the normal and the orientation. The default orientation is that the normals for your internal sidesets point _inward_ to the cavity (see the picture below).
![IMG_D5CD61EEEE41-1](https://user-images.githubusercontent.com/22062133/118140634-c5990080-b3c5-11eb-8ff4-5d078f1f4bc8.jpeg)

We probably need to improve the documentation here a bit 😬

**Reply by makeclean** [reply]
Thanks @loganharbour I believe the sideset convention above is honored by my geometry? I should point out I don't have *any* internal sidesets, the region between the cylinders and the start of the hollow brick, is not meshed.

![Screenshot from 2021-05-14 07-59-36](https://user-images.githubusercontent.com/2439754/118233994-a021fa80-b48a-11eb-9d0b-7054c1d1e830.png)
![Screenshot from 2021-05-14 07-59-53](https://user-images.githubusercontent.com/2439754/118233999-a1532780-b48a-11eb-8713-ed41731cc443.png)

Or are you saying, my normals all need reversing?

I think this requirement may be a little impossible to achieve in some models, given what Cubit does during imprinting and merging, surfaces that become merged can have inconsistent normals in this scheme and I think this will be true of the sidesets they belong to. E.g. two cubes that have an adjacent shared face will have a normal that points in a particular direction, the surface that is shared has a property sense_forward which can be FORWARD or REVERSE. I have no control over this if want to produce a consistent mesh.

**Reply by loganharbour** [reply]
> I should point out I don't have any internal sidesets, the region between the cylinders and the start of the hollow brick, is not meshed.

Ahh, this is actually a bigger issue. With the initial ray tracing capability, we're required to traverse through the mesh even when tracing from a boundary to another boundary. There are plans in the future to look into tracing by checking all possible boundary surfaces instead of requiring a mesh, but we unfortunately don't have time right now to do this. 

If I recall correctly, @helen-brooks is looking into something that is like this (boundary-to-boundary tracing), but I cannot recall if it is using MOOSE or just libMesh.

**Reply by helen-brooks** [reply]
@loganharbour The project you are referring to is using libMesh to do surface ray-tracing within DAGMC, which @makeclean is aware of since it was his idea. Recently haven't had much time to work on this, though if it's something you or the MOOSE community would be interested in, it might be an argument to put in more resources.
